README.md
# Trends

[![Travis-CI Build Status](https://travis-ci.org/jfisher-usgs/Trends.svg?branch=master)](https://travis-ci.org/jfisher-usgs/Trends)

This [R](http://www.r-project.org/ "R") package is for identifying trends in data from multiple observation sites in a monitoring network.
A parametric survival regression model is used to fit the observed data, both censored and uncensored.
The set of standards used for coding this package is documented in
[Google's R Style Guide](https://google.github.io/styleguide/Rguide.xml "Google's R Style Guide").

## Install

If R is not already installed on your computer, download and install the latest binary distribution from
[CRAN](http://cran.r-project.org/ "The Comprehensive R Archive Network").
Next, open an R session and install the required packages using the following commands:

```r
repos <- c("https://jfisher-usgs.github.io/R", getOption("repos"))
install.packages("Trends", repos = repos, dependencies = TRUE)
```

Support for merging PDF files into a new file requires [PDFtk Server](http://www.pdflabs.com/tools/pdftk-server/ "pdftk"),
a cross-platform command-line tool for working with PDFs; download and install.

## Run

Load **Trends** in the current R session:

```r
library(Trends)
```

See help documentation:

```r
help(package = "Trends")
```

## Bugs

Please consider reporting bugs and asking questions on the
[Issues page](https://github.com/jfisher-usgs/Trends/issues "Issues page").

## Disclaimer

This software is in the public domain because it contains materials that originally came from the U.S. Geological Survey, an agency of the United States Department of Interior.
For more information, see the [official USGS copyright policy](http://www.usgs.gov/visual-id/credit_usgs.html#copyright/ "official USGS copyright policy").

Although this software program has been used by the U.S. Geological Survey (USGS), no warranty, expressed or implied,
is made by the USGS or the U.S. Government as to the accuracy and functioning of the program and related program material nor shall the fact of distribution constitute any such warranty,
and no responsibility is assumed by the USGS in connection therewith.

This software is provided "AS IS."

[![CC0](http://i.creativecommons.org/p/zero/1.0/88x31.png)](http://creativecommons.org/publicdomain/zero/1.0/)


NEWS
Changes in Version 1.1.0

  o  Change to in-source documentation using roxygen2 package.

Changes in Version 1.0.0

  o  Change repository for package installation to https://jfisher-usgs.github.io/R

  o  Minor changes for compiling with R version 3.3.0.

  o  Change format for package version numbering from #.#-# to #.#.#

Changes in Version 0.2-1

  o  Additional explanatory variables (such as seasonality and water levels) may be added as covariants in the trend model.

Changes in Version 0.2-0

  o  This version represents a major change in the package and breaks compatibility with older versions.
     The impetus for this change was a need to model interval-censored data.
     A parametric survival regression model is fit to the observed data, both censored and uncensored.

Changes in Version 0.1-5

  o  Fixed bug in RunTrendAnalysis function that dropped matrix class for spatial data frames with a single coordinate value.

  o  Fixed bug in RunTrendAnalysis function that tried to compute trend when summary statistics returned from NADA::cenfit function were incomplete.

  o  Added a package CITATION file and a LaTeX bibliography source file.

  o  Added argument in MergePDFs function to remove individual files after merge is completed.

Changes in Version 0.1-4

  o  Added a package vignette describing 2014 water quality data processing.

  o  Summary statistics are written to a shapefile.

  o  Added new functions for reading configuration files (ReadParConfig and ReadPlotConfig) and a geo-referenced site location file (ReadSiteLocations).

  o  The 'Min', 'Max', 'Start_date', and 'End_date' fields have been removed from the plot configuration files.
     Users need to update their input files to account for this change.

  o  Graphics now opened in a platform-independent way using the grDevices::dev.new function.

  o  Many of the function names were updated using a more explicit naming convention.

  o  The tcltk package is no longer required. Removed Tcl/Tk dialog boxes for file and directory selection.
     Default directory arguments are the working directory and file path names must be passed as arguments.

  o  Fixed calculation of the number of samples above the reporting limit (n_above_rl); previous value did not include the number of estimated (E) samples.

  o  Plot legends show "p-value < 0.001"; previously these labels were shown as "p-value = 0.000".

  o  Nonparametric regression is prevented when NADA::cenfit function results in warnings.

  o  Warn if site ids in configuration files do not match ids in the data file.

  o  Added a new function MergePDFs, combines PDF files into a new file.


R/TMP.R

################################################################################

.MergePDFs <- function(path, retain.files=FALSE) {
  if (Sys.which("pdftk") == "")
    stop("pdftk not found, check that PDFtk Server is installed")

  pdfs <- list.files(path, pattern=".pdf$")
  if (length(pdfs) == 0 || pdfs == "")
    stop("path does not exist or input files are missing")

  out.pdf <- file.path(dirname(path), paste0(basename(path), ".pdf"))
  if(file.exists(out.pdf))
    file.remove(out.pdf)

  tmp.txt <- tempfile(fileext=".txt")
  tmp.bat <- tempfile(fileext=".bat")
  tmp.pdf <- tempfile(fileext=".pdf")

  cmd <- paste("cd", shQuote(path))

  cat("", file=tmp.bat)
  Sys.chmod(tmp.bat, mode="755")

  npages <- NULL
  for (i in pdfs) {
    cmd[2] <- paste("pdftk", shQuote(i), "dump_data output", shQuote(tmp.txt))
    cat(cmd, file=tmp.bat, sep="\n")
    system(command=tmp.bat, show.output.on.console=FALSE)
    txt <- scan(tmp.txt, what=character(), quiet=TRUE)
    npages <- c(npages, as.integer(txt[which(txt == "NumberOfPages:") + 1L]))
  }
  pages <- cumsum(npages) - (npages - 1L)

  cmd[2] <- paste("pdftk *.pdf cat output", shQuote(tmp.pdf))
  cat(cmd, file=tmp.bat, sep="\n")
  system(command=tmp.bat, show.output.on.console=FALSE)

  cmd[2] <- paste("pdftk", shQuote(tmp.pdf), "dump_data output",
                  shQuote(tmp.txt))
  cat(cmd, file=tmp.bat, sep="\n")
  system(command=tmp.bat, show.output.on.console=FALSE)

  bookmarks <- sub("\\.pdf$", "", pdfs)
  FUN <- function(i) {
    bookmark <- c("BookmarkBegin", paste("BookmarkTitle:", bookmarks[i]),
                  "BookmarkLevel: 1", paste("BookmarkPageNumber:", pages[i]))
    return(paste(bookmark, collapse="\n"))
  }
  bookmarks <- vapply(seq_along(bookmarks), FUN, "")
  cat(bookmarks, file=tmp.txt, sep="\n", append=TRUE)
  cmd[2] <- paste("pdftk", shQuote(tmp.pdf), "update_info", shQuote(tmp.txt),
                  "output", shQuote(out.pdf))
  cat(cmd, file=tmp.bat, sep="\n")
  system(command=tmp.bat, show.output.on.console=FALSE)

  unlink(tmp.txt)
  unlink(tmp.bat)
  unlink(tmp.pdf)
  if (!retain.files) {
    unlink(file.path(path, "*.pdf"))
    if (length(list.files(path)) == 0)
      unlink(path, recursive=TRUE, force=TRUE)
  }

  invisible(out.pdf)
}

################################################################################

.OpenDevice <- function(path, id, graphics.type, w=8.5, h=11) {
  if (missing(graphics.type) || !graphics.type %in% c("pdf", "eps")) {
    dev.new(width=w, height=h)
  } else {
    graphics.off()
    file <- file.path(path, paste(id, graphics.type, sep="."))
    if (file.access(file) == 0) {
      warning(paste("file already exists and will be overwritten:", file))
    }
    if (graphics.type == "pdf") {
      pdf(file=file, width=w, height=h, version="1.6", colormodel="cmyk")
    } else if (graphics.type == "eps") {
      postscript(file=file, width=w, height=h, horizontal=FALSE, paper="letter")
    }
  }
  par(mfrow=c(4, 1), omi=c(4, 3.5, 6, 4.5) * 0.166667)
}

################################################################################

.ProcessRawData <- function(raw.data, parameters, detection.limits=NULL,
                            date.fmt="%Y-%m-%d") {

  raw.data$Date <- as.Date(raw.data$Date, format=date.fmt)
  raw.data <- raw.data[!is.na(raw.data$Date), ]

  parameters$pch <- as.integer(parameters$pch)
  par.names <- make.names(parameters$Parameter)
  par.names <- sort(par.names[par.names %in% colnames(raw.data)])

  detection.limits$Date <- as.Date(detection.limits$Date, format=date.fmt)
  idx <- which(names(detection.limits) %in% "Date")
  detection.limits[, -idx] <- apply(detection.limits[, -idx], 2, as.numeric)

  lst <- list()
  for (i in seq_along(par.names)) {
    nam <- par.names[i]
    is.rec <- !is.na(raw.data[[nam]]) & raw.data[[nam]] != ""
    if (all(!is.rec))
      next

    d <- raw.data[is.rec, c("Site_name", "Site_id", "Date")]
    d <- data.frame(d, code=NA, conc=NA, sd=NA, dl=NA, t1=NA, t2=NA,
                    is.exact=NA, is.left=NA, is.interval=NA)
    d$Site_id   <- as.factor(d$Site_id)
    d$Site_name <- as.factor(d$Site_name)

    d$conc <- as.character(raw.data[is.rec, par.names[i]])
    d$code <- substr(d$conc, 1, 1)
    d$code[!d$code %in% c("<", "E", "V", "U")] <- ""
    d$code <- as.factor(d$code)
    is.code <- d$code != ""
    d$conc[is.code] <- substr(d$conc[is.code], 2, nchar(d$conc[is.code]))
    d$conc <- as.numeric(d$conc)
    d$conc[d$code %in% c("V", "U")] <- NA

    p <- parameters[match(nam, make.names(parameters$Parameter)), , drop=TRUE]
    sd.col <- p$sd
    idx <- ifelse(!is.na(sd.col), match(sd.col, colnames(raw.data)), NA)
    d$sd <- if (is.na(idx)) NA else as.numeric(raw.data[is.rec, idx])

    if (nam %in% colnames(detection.limits)) {
      dl <- detection.limits[!is.na(detection.limits[[nam]]), c("Date", nam)]
      for (id in levels(d$Site_id)) {
        idxs <- which(d$Site_id == id)
        breaks <- findInterval(as.numeric(dl$Date), as.numeric(d$Date[idxs]))
        breaks <- unique(c(breaks, length(idxs) + 1L))
        d$dl[idxs] <- dl[as.integer(cut(seq_along(idxs), breaks)), nam]
      }
    }

    is.left <- d$code %in% "<"
    is.exact <- !is.left & (is.na(d$sd) | is.na(d$dl))
    d$t1[is.exact] <- d$conc[is.exact]
    d$t2[is.exact] <- d$conc[is.exact]
    d$t2[is.left]  <- d$conc[is.left]
    lower.conc <- d$conc - 3 * d$sd
    upper.conc <- d$conc + 3 * d$sd

    are.left <- which(!is.exact & upper.conc <= d$dl)
    d$t2[are.left] <- d$dl[are.left]
    are.left <- which(!is.exact & lower.conc <= d$dl & upper.conc > d$dl)
    d$t2[are.left] <- upper.conc[are.left]

    are.interval <- which(!is.exact & lower.conc > d$dl)
    d$t1[are.interval] <- lower.conc[are.interval]
    d$t2[are.interval] <- upper.conc[are.interval]

    d$t2[which(d$t2 <= 0)] <- NA  # TODO: zero is invalid for 'lognormal' distribution
    d$t1[which(d$t1 <= 0 | is.na(d$t2))] <- NA

    is.t1 <- !is.na(d$t1)
    is.t2 <- !is.na(d$t2)
    d$is.exact    <-  is.t1 & is.t2 & d$t1 == d$t2
    d$is.left     <- !is.t1 & is.t2
    d$is.interval <-  is.t1 & is.t2 & d$t1 != d$t2
    d <- d[order(d$Site_name, d$Date), ]
    attributes(d) <- c(attributes(d), p[c("Parameter", "Name", "Units", "sd")])

    lst[[p$Parameter]] <- d
  }

  return(lst)
}

################################################################################

.ProcessConfig <- function(config, processed.data) {
  ids <- unique(unlist(lapply(processed.data, function(i) levels(i$Site_id))))
  config <- config[config$Site_id %in% ids, ]
  FUN <- function(i) {
    d <- config[i, , drop=FALSE]
    p <- strsplit(d$Parameters, ",")[[1]]
    p <- unique(sub("^[[:space:]]*(.*?)[[:space:]]*$", "\\1", p, perl=TRUE))
    d <- data.frame(d, Parameter=p, rec=i, row.names=NULL,
                    stringsAsFactors=FALSE)
    d$Parameters <- NULL
    return(d)
  }
  d <- do.call(rbind, lapply(seq_len(nrow(config)), FUN))
  d <- d[d$Parameter %in% names(processed.data), ]
  d <- d[order(as.integer(factor(d$Site_id, levels=unique(d$Site_id)))), ]

  if (is.null(d$Axis_title))
    d$Axis_title <- NA

  return(d)
}

################################################################################

.GetModelInfo <- function(model) {
  x <- c()
  x[c("c1", "c2")] <- model$coefficients
  x["scale"] <- model$scale
  x["p"] <- 1 - pchisq(2 * diff(model$loglik), sum(model$df) - model$idf)
  x["slope"] <- 100 * (exp(model$coefficients[2]) - 1) * 365.242  # in percent change per year
  return(x)
}

################################################################################

.CreateDir <- function(path, id, graphics.type) {
  dir.create(path=path, showWarnings=FALSE, recursive=TRUE)
  if (!graphics.type %in% c("pdf", "eps"))
    return()
  id.path <- file.path(path, id)
  if (file.exists(id.path))
    unlink(file.path(id.path, paste0("*.", graphics.type)))
  dir.create(path=id.path, showWarnings=FALSE, recursive=TRUE)
  return(id.path)
}

################################################################################

.RunAnalysis <- function(processed.data, processed.config, sdate=NA, edate=NA,
                         graphics.type="pdf", merge.pdfs=TRUE, path=getwd(),
                         id=format(Sys.time(), "%Y%m%d%H%M%S"),
                         site.locations=NULL) {

  if(inherits(sdate <- try(as.Date(sdate)), "ty-error"))
    sdate <- NA
  if(inherits(edate <- try(as.Date(edate)), "ty-error"))
    edate <- NA

  obs <- list()
  models <- list()

  d <- processed.config[, c("Site_id", "Site_name", "Parameter")]
  stats <- data.frame(d, "sdate"=sdate, "edate"=edate, "n"=NA, "nmissing"=NA,
                      "nexact"=NA, "nleft"=NA, "ninterval"=NA, "nbelow.rl"=NA,
                      "min"=NA, "max"=NA, "median"=NA, "mean"=NA, "sd"=NA,
                      "0.95LCL"=NA, "0.95UCL"=NA, "c1"=NA, "c2"=NA, "scale"=NA,
                      "p"=NA, "slope"=NA, "trend"=NA, check.names=FALSE)

  for (i in seq_len(nrow(processed.config))) {
    d <- processed.data[[processed.config[i, "Parameter"]]]
    d <- d[d$Site_id == processed.config[i, "Site_id"], ]

    date1 <- if (is.na(sdate)) min(d$Date) else sdate
    date2 <- if (is.na(edate)) max(d$Date) else edate
    d <- d[d$Date >= date1 & d$Date <= date2, ]

    d$surv <- Surv(time=d$t1, time2=d$t2, type="interval2")
    is.missing <- is.na(d$surv)

    obs[[i]] <-  d

    stats[i, c("n", "nmissing")] <- c(nrow(d), sum(is.missing))
    vars <- c("nexact", "nleft", "ninterval")
    stats[i, vars] <- c(sum(d$is.exact), sum(d$is.left), sum(d$is.interval))
    stats[i, "nbelow.rl"] <- sum(d$code == "<")

    stats[i, c("min", "max")] <- c(min(d$t1), max(d$t2))

    model <- suppressWarnings(survreg(surv ~ Date, data=d, dist="lognormal",
                                      control=survreg.control(maxiter=100)))
    models[[i]] <- model
    model.info <- .GetModelInfo(model)

    stats[i, names(model.info)] <- model.info
    p <- model.info["p"]
    slope <- model.info["slope"]
    is.trend <- !anyNA(c(p, slope)) & p <= 0.05
    stats[i, "trend"] <- ifelse(is.trend, ifelse(slope > 0, "+", "-"), "none")

    if (any(d$is.interval)) {
      fit <- summary(survfit(d$surv ~ 1))$table
      vars <- c("median", "0.95LCL", "0.95UCL")
      stats[i, vars] <- fit[vars]
    } else if (any(d$is.left)) {
      fit <- cenfit(d$t2[!is.missing], d$is.left[!is.missing])
      vars <- c("mean", "0.95LCL", "0.95UCL")
      stats[i, vars] <- suppressWarnings(mean(fit)[vars])
      stats[i, c("median", "sd")] <- suppressWarnings(c(median(fit), sd(fit)))
    } else {
      t1 <- na.omit(d$t1)
      stats[i, c("median", "mean", "sd")] <- c(median(t1), mean(t1), sd(t1))
    }
  }

  id.path <- .CreateDir(path, id, graphics.type)

  plot.count <- list()
  for (i in seq_len(nrow(processed.config))) {
    site.id <- processed.config$Site_id[i]
    site.name <- processed.config$Site_name[i]

    if (is.null(plot.count[[site.id]]))
      plot.count[[site.id]] <- 0L
    if ((plot.count[[site.id]] + 4L) %% 4L == 0L) {
      letter <- LETTERS[(plot.count[[site.id]] + 4L) %/% 4L]
      .OpenDevice(id.path, paste0(site.name, "_", letter), graphics.type)
      main <- paste0(site.name, " (", site.id, ")")
    } else {
      main <- NULL
    }
    plot.count[[site.id]] <- plot.count[[site.id]] + 1L

    a <- attributes(processed.data[[processed.config$Parameter[i]]])
    ylab <- processed.config$Axis_title[i]
    if (is.na(ylab))
      ylab <- ifelse(is.na(a$Units), a$Name, paste0(a$Name, ", in ", a$Units))
    xlim <- if (inherits(c(sdate, edate), "Date")) c(sdate, edate) else NULL
    .DrawPlot(obs[[i]], models[[i]], xlim=xlim, main=main, ylab=ylab)
  }
  if (graphics.type %in% c("pdf", "eps"))
    graphics.off()
  if (graphics.type == "pdf" && merge.pdfs)
    .MergePDFs(id.path)

  file <- file.path(path, paste0(id, ".tsv"))
  write.table(stats, file=file, quote=FALSE, sep="\t", row.names=FALSE)
  if (inherits(site.locations, "SpatialPointsDataFrame")) {
    idxs <- match(stats$Site_id, site.locations@data$Site_id)
    if (anyNA(idxs)) {
      stop("site id(s) not found in 'spatial.locations'")
    } else {
      coords <- site.locations@coords[idxs, , drop=FALSE]
      crs <- site.locations@proj4string
      obj <- SpatialPointsDataFrame(coords, stats, proj4string=crs)
      writeOGR(obj, path, id, "ESRI Shapefile", check_exists=TRUE,
               overwrite_layer=TRUE)
    }
  }

  invisible(stats)
}

################################################################################

.PlotObs <- function(processed.data, processed.config, sdate=NA, edate=NA,
                     graphics.type="pdf", merge.pdfs=TRUE, path=getwd(),
                     id=format(Sys.time(), "%Y%m%d%H%M%S")) {

  if(inherits(sdate <- try(as.Date(sdate)), "ty-error"))
    sdate <- NA
  if(inherits(edate <- try(as.Date(edate)), "ty-error"))
    edate <- NA

  id.path <- .CreateDir(path, id, graphics.type)

  plot.count <- list()
  rec <- processed.config$rec
  for (i in sort(unique(rec))) {

    idx <- match(i, rec)
    site.id   <- processed.config[idx, "Site_id"]
    site.name <- processed.config[idx, "Site_name"]

    pars <- processed.config[rec == i, "Parameter"]
    lst <- lapply(pars, function(i) processed.data[[i]])
    names(lst) <- pars

    FUN <- function(i) {
      return(data.frame(lst[[i]], Name=attr(lst[[i]], "Name"),
                        stringsAsFactors=FALSE))
    }
    d <- do.call(rbind, lapply(seq_along(lst), FUN))
    d <- d[d$Site_id == site.id, ]

    date1 <- if (is.na(sdate)) min(d$Date) else sdate
    date2 <- if (is.na(edate)) max(d$Date) else edate
    d <- d[d$Date >= date1 & d$Date <= date2, ]

    if (is.null(plot.count[[site.id]]))
      plot.count[[site.id]] <- 0L
    if ((plot.count[[site.id]] + 4L) %% 4L == 0L) {
      letter <- LETTERS[(plot.count[[site.id]] + 4L) %/% 4L]
      .OpenDevice(id.path, paste0(site.name, "_", letter), graphics.type)
      main <- paste0(site.name, " (", site.id, ")")
    } else {
      main <- NULL
    }
    plot.count[[site.id]] <- plot.count[[site.id]] + 1L

    ylab <- processed.config[idx, "Axis_title"]
    if (is.na(ylab))
      ylab <- "Unknown"
    xlim <- if (inherits(c(sdate, edate), "Date")) c(sdate, edate) else NULL

    .DrawPlot(d, xlim=xlim, main=main, ylab=ylab)
  }
  if (graphics.type %in% c("pdf", "eps"))
    graphics.off()
  if (graphics.type == "pdf" && merge.pdfs)
    .MergePDFs(id.path)
}

################################################################################

.DrawPlot <- function(obj, model, xlim=NULL, ylim=NULL, main=NULL, ylab=NULL) {

  obj$t1[is.na(obj$t1) & obj$is.left] <- 0
  if (!missing(model))
    obj <- obj[!is.na(obj$surv), ]

  xran <- extendrange(obj$Date, f=0.02)
  if (inherits(xlim, "Date")) {
    xlim[1] <- if (is.na(xlim[1])) xran[1] else xlim[1]
    xlim[2] <- if (is.na(xlim[2])) xran[2] else xlim[2]
  } else {
    xlim <- xran
  }
  if (is.null(ylim)) {
    ylim <- c(min(obj$t1), max(obj$t2, na.rm=TRUE))
    ylim <- extendrange(pretty(ylim), f=0.06)
  }

  par(mar=c(1, 3, 2, 0.5) + 0.1, mgp=c(2, 0.5, 0))
  plot(NA, xlim=xlim, ylim=ylim, xaxt="n", yaxt="n", xaxs="i", yaxs="i",
       xlab="Date", ylab=ylab, type="n", main=main, frame.plot=FALSE)

  if (missing(model)) {
    cols <- c("#F02311", "#C3FF68", "#107FC9", "#BE80FF", "#050505")
    obj$col <- cols[as.integer(factor(obj$Name, levels=unique(obj$Name)))]
  } else {
    x <- seq(xlim[1], xlim[2], "days")
    y <- predict(model, list(Date=x), type="quantile", p=c(0.1, 0.9, 0.5))
    polygon(c(x, rev(x)), c(y[, 1], rev(y[, 2])), col="#FFFFD5", border=NA)
    lines(x, y[, 3], lty=1, lwd=1, col="#F02311")
    obj$col <- "#107FC9"
  }

  lwd <- 0.5 * (96 / (6 * 12))
  tcl <- 0.50 / (6 * par("csi"))

  is.int <- obj$is.left | obj$is.interval
  if (any(is.int)) {
    o <- obj[is.int, ]
    suppressWarnings(arrows(x0=o$Date, y0=o$t1, y1=o$t2, length=0.015, angle=90,
                            code=3, col=o$col, lwd=1))
  }
  if (any(obj$is.exact)) {
    o <- obj[obj$is.exact, ]
    points(x=o$Date, y=o$t2, pch=20, col=o$col)
  }

  at <- pretty(xlim, n=10)
  axis.Date(1, xlim, at, tcl=tcl, lwd=-1, lwd.ticks=lwd)
  axis.Date(3, xlim, at, tcl=tcl, lwd=-1, lwd.ticks=lwd, labels=FALSE)
  axis(2, tcl=tcl, lwd=-1, lwd.ticks=lwd)
  axis(4, tcl=tcl, lwd=-1, lwd.ticks=lwd, labels=FALSE)

  box(lwd=lwd)

  inset <- c(0.02, 0.02 * do.call("/", as.list(par("pin"))))
  if (missing(model)) {
    o <- obj[!duplicated(obj$Name), ]
    suppressWarnings(legend("topleft", o$Name, fill=o$col, border=NA, xpd=NA,
                            bg="#FFFFFFBB", box.lwd=lwd, inset=inset))
  } else {
    p <- .GetModelInfo(model)["p"]
    if (is.na(p))
      return()
    p <- ifelse(p < 0.001, "p < 0.001", paste("p =", sprintf("%.3f", p)))
    txt <- paste0("Regression, ", p)
    suppressWarnings(legend("topleft", txt, lty=1, col="#F02311", xpd=NA,
                            bg="#FFFFFFBB", box.lwd=lwd, inset=inset))
  }
}

################################################################################

.tmp <- function() {  # TODO: move to vignette

  require(NADA); require(rgdal)


  path.in <- system.file("extdata", "SIR2014", package = "Trends")
  read.args <- list(header = TRUE, sep = "\t", colClasses = "character", na.strings = "",
                    fill = TRUE, strip.white = TRUE, comment.char = "", flush = TRUE,
                    stringsAsFactors = FALSE)

  file <- file.path(path.in, "Raw_Data.tsv")
  raw.data <- do.call(read.table, c(list(file), read.args))

  file <- file.path(path.in, "Parameters.tsv")
  parameters <- do.call(read.table, c(list(file), read.args))

  file <- file.path(path.in, "Detection_Limits.tsv")
  detection.limits <- do.call(read.table, c(list(file), read.args))

  processed.data <- .ProcessRawData(raw.data, parameters, detection.limits,
                                    date.fmt = "%m/%d/%Y")

  site.locations <- readOGR(path.in, layer = "Site_Locations", verbose = FALSE)

  path.out <- "C:/Users/jfisher/Desktop/Trends"


  file <- file.path(path.in, "Config_Plots.tsv")
  config <- do.call(read.table, c(list(file), read.args))
  processed.config <- .ProcessConfig(config, processed.data)
  .PlotObs(processed.data, processed.config, id = "Data_1960-2012",
           sdate = "1960-01-01", edate = "2012-12-31", path = path.out)


  file <- file.path(path.in, "Config_Plots.tsv")
  config <- do.call(read.table, c(list(file), read.args))
  processed.config <- .ProcessConfig(config, processed.data)
  .PlotObs(processed.data, processed.config, id = "Data_1989-2012",
           sdate = "1989-01-01", edate = "2012-12-31", path = path.out)


  file <- file.path(path.in, "Config_Plots_Field.tsv")
  config <- do.call(read.table, c(list(file), read.args))
  processed.config <- .ProcessConfig(config, processed.data)
  .PlotObs(processed.data, processed.config, id = "Data_1960-2012_Field",
           sdate = "1960-01-01", edate = "2012-12-31", path = path.out)


  file <- file.path(path.in, "Config_Plots_Field.tsv")
  config <- do.call(read.table, c(list(file), read.args))
  processed.config <- .ProcessConfig(config, processed.data)
  .PlotObs(processed.data, processed.config, id = "Data_1989-2012_Field",
           sdate = "1989-01-01", edate = "2012-12-31", path = path.out)


  file <- file.path(path.in, "Config_Cen.tsv")
  config <- do.call(read.table, c(list(file), read.args))
  processed.config <- .ProcessConfig(config, processed.data)
  stats <- .RunAnalysis(processed.data, processed.config, id = "Stats_1989-2012_Cen",
                        sdate = "1989-01-01", edate = "2012-12-31",
                        path = path.out, site.locations = site.locations)


  file <- file.path(path.in, "Config_Uncen.tsv")
  config <- do.call(read.table, c(list(file), read.args))
  processed.config <- .ProcessConfig(config, processed.data)
  stats <- .RunAnalysis(processed.data, processed.config, id = "Stats_1960-2012_Uncen",
                        sdate = "1960-01-01", edate = "2012-12-31",
                        path = path.out, site.locations = site.locations)


  file <- file.path(path.in, "Config_Uncen.tsv")
  config <- do.call(read.table, c(list(file), read.args))
  processed.config <- .ProcessConfig(config, processed.data)
  stats <- .RunAnalysis(processed.data, processed.config, id = "Stats_1989-2012_Uncen",
                        sdate = "1989-01-01", edate = "2012-12-31",
                        path = path.out, site.locations = site.locations)


  file <- file.path(path.in, "Config_Uncen_Field.tsv")
  config <- do.call(read.table, c(list(file), read.args))
  processed.config <- .ProcessConfig(config, processed.data)
  stats <- .RunAnalysis(processed.data, processed.config, id = "Stats_1960-2012_Uncen_Field",
                        sdate = "1960-01-01", edate = "2012-12-31",
                        path = path.out, site.locations = site.locations)


  file <- file.path(path.in, "Config_Uncen_Field.tsv")
  config <- do.call(read.table, c(list(file), read.args))
  processed.config <- .ProcessConfig(config, processed.data)
  stats <- .RunAnalysis(processed.data, processed.config, id = "Stats_1989-2012_Uncen_Field",
                        sdate = "1989-01-01", edate = "2012-12-31",
                        path = path.out, site.locations = site.locations)


  file <- file.path(path.in, "Config_Cen_VOC.tsv")
  config <- do.call(read.table, c(list(file), read.args))
  processed.config <- .ProcessConfig(config, processed.data)
  stats <- .RunAnalysis(processed.data, processed.config, id = "Stats_1987-2012_Cen_VOC",
                        sdate = "1987-01-01", edate = "2012-12-31",
                        path = path.out, site.locations = site.locations)


  file <- file.path(path.in, "Config_Uncen_VOC.tsv")
  config <- do.call(read.table, c(list(file), read.args))
  processed.config <- .ProcessConfig(config, processed.data)
  stats <- .RunAnalysis(processed.data, processed.config, id = "Stats_1987-2012_Uncen_VOC",
                        sdate = "1987-01-01", edate = "2012-12-31",
                        path = path.out, site.locations = site.locations)


  file <- file.path(path.in, "Config_RADS.tsv")
  config <- do.call(read.table, c(list(file), read.args))
  processed.config <- .ProcessConfig(config, processed.data)
  stats <- .RunAnalysis(processed.data, processed.config, id = "Stats_1981-2012_RAD",
                        sdate = "1981-01-01", edate = "2012-12-31",
                        path = path.out, site.locations = site.locations)





}


